libcudacxx/include/cuda/std/__atomic/functions/cuda_ptx_derived.h

```diff
@@ -33,6 +31,244 @@ _LIBCUDACXX_BEGIN_NAMESPACE_STD
+_CCCL_DEVICE _Type __cuda_atomic_fetch_update(_Type* __ptr, const _Fn& __op, _Order, _Operand, _Sco)
+{
+  constexpr uint64_t __alignmask = (sizeof(uint32_t) - 1);
+  constexpr uint32_t __sizemask  = (1 << (sizeof(_Type) * 8)) - 1;
+  uint32_t* __aligned            = (uint32_t*) ((intptr_t) __ptr & (~__alignmask));
+  const uint8_t __offset         = uint8_t((intptr_t) __ptr & __alignmask) * 8;
+  const uint32_t __valueMask     = __sizemask << __offset;
+  const uint32_t __windowMask    = ~__valueMask;
+
+  // 8/16b fetch update is similar to CAS implementation, but compresses the logic for recalculating the operand
+  // __old = __window[0:32] where [__cmp] resides within some offset.
+  uint32_t __old;
+  NV_IF_TARGET(
+    NV_PROVIDES_SM_70,
+    (__cuda_atomic_load(
+       __aligned, __old, __atomic_cuda_relaxed{}, __atomic_cuda_operand_b32{}, _Sco{}, __atomic_cuda_mmio_disable{});),
+    (__cuda_atomic_load(
+       __aligned, __old, __atomic_cuda_volatile{}, __atomic_cuda_operand_b32{}, _Sco{}, __atomic_cuda_mmio_disable{});))
+
+  // Reemit CAS instructions until we succeed
+  while (1)
+  {
+    // Calculate new desired value from last fetched __old
+    // Use of the value mask is required due to the possibility of overflow when ops are widened. Possible compiler bug?
+    const uint32_t __attempt =
+      ((static_cast<uint32_t>(__op(static_cast<_Type>(__old >> __offset))) << __offset) & __valueMask)
+      | (__old & __windowMask);
+
+    if (__cuda_atomic_compare_exchange(
+          __aligned, __old, __old, __attempt, _Order{}, __atomic_cuda_operand_b32{}, _Sco{}))
+    {
+      // CAS was successful
+      return static_cast<_Type>(__old >> __offset);
+    }
+  }
+}
+
+// Optimized fetch_update CAS loop with op determined after first load reducing waste.
+template <class _Type,
+          class _Fn,
+          class _Order,
+          class _Operand,
+          class _Sco,
+          __cuda_atomic_enable_native_bitwise<_Operand> = 0>
+_CCCL_DEVICE _Type __cuda_atomic_fetch_update(_Type* __ptr, const _Fn& __op, _Order, _Operand, _Sco)
+{
+  _Type __expected = 0;
+  NV_IF_TARGET(
+    NV_PROVIDES_SM_70,
+    (__cuda_atomic_load(
+       __ptr, __expected, __atomic_cuda_relaxed{}, __atomic_cuda_operand_b32{}, _Sco{}, __atomic_cuda_mmio_disable{});),
+    (__cuda_atomic_load(
+       __ptr, __expected, __atomic_cuda_volatile{}, __atomic_cuda_operand_b32{}, _Sco{}, __atomic_cuda_mmio_disable{});))
+
+  _Type __desired = __op(__expected);
+  while (!__cuda_atomic_compare_exchange(__ptr, __expected, __expected, __desired, _Order{}, _Operand{}, _Sco{}))
+  {
+    __desired = __op(__expected);
+  }
+  return __expected;
+}
+
+template <class _Type, class _Order, class _Operand, class _Sco, __cuda_atomic_enable_non_native_bitwise<_Operand> = 0>
+static inline _CCCL_DEVICE void
+__cuda_atomic_store(_Type* __ptr, _Type __val, _Order, _Operand, _Sco, __atomic_cuda_mmio_disable)
+{
+  // Store requires cas on 8/16b types
```

Comment:
Why does store require cas on 8 and 16b types?

PTX atomic store (`st.relaxed`,`st.release`) supports `.b8`,`.u8`,`.s8`,`.b16`,`.u16`,`.s16`:
https://docs.nvidia.com/cuda/parallel-thread-execution/index.html#data-movement-and-conversion-instructions-st

Reply:
I've updated the PR to use 16b ld/st. To make use of 8b we'd need to dispatch to assembly that uses a widened proxy for output/input. That will require adjustments that I don't want to add into this PR. We can do it later.

Reply:
> More poignantly, there is no PTX constraint for 8b values.

Right, libcu++ needs to zero extend the 8-bit value to a 16 bit value, pass the 16-bit value to PTX, and then directly pass that to the 8byte load/store ops.